### ISSUE TITLE
Incorrect handling of EOS tokens in DataCollatorForLanguageModeling when pad_token is set to eos_token

### ISSUE THREAD

**Issue by sapulidiularlidi** [UKPLab/sentence-transformers#1952]
System Info
Doesn't seem to be version specific, but:

transformers version: 4.29.2
Platform: Linux-5.15.107+-x86_64-with-glibc2.31
Python version: 3.10.11
Huggingface_hub version: 0.14.1
Safetensors version: not installed
PyTorch version (GPU?): 2.0.1+cu118 (False)
Tensorflow version (GPU?): 2.12.0 (False)
Flax version (CPU?/GPU?/TPU?): 0.6.9 (cpu)
Jax version: 0.4.8
JaxLib version: 0.4.7
Using GPU in script?: No
Using distributed or parallel set-up in script?: No
Who can help?
@sgugger I think?

Information
 The official example scripts
 My own modified scripts
Tasks
 An officially supported task in the examples folder (such as GLUE/SQuAD, ...)
 My own task or dataset (give details below)
Reproduction
When using DataCollatorForLanguageModeling for CLM with pad_token set to eos_token as it's shown [here](https://huggingface.co/docs/transformers/tasks/language_modeling) - all EOS tokens in the labels are overwritten with -100, instead of just the ones used for padding.

In colab [here](https://colab.research.google.com/drive/13JsslKctbc9JEWbsJEJ6xru4zm6QxmLt?usp=sharing).

Prepare sentences that should be included in a single batch with explicit EOS tokens. (here for GPT-2)
sentences = ["Short sentence<|endoftext|>", "Now that's a longer sentence<|endoftext|>"]
Tokenize them.
tokenizer = AutoTokenizer.from_pretrained('gpt2')

tokenized_sentences = [tokenizer(sentence) for sentence in sentences]
Collate them with DataCollatorForLanguageModeling.
tokenizer.pad_token = tokenizer.eos_token # 50256
data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)
batch = data_collator(tokenized_sentences)
Batch is:

{'input_ids': tensor([[16438,  6827, 50256, 50256, 50256, 50256, 50256], 
                      [3844, 326, 338, 257, 2392, 6827, 50256]]), 
'attention_mask': tensor([[1, 1, 1, 0, 0, 0, 0], 
                          [1, 1, 1, 1, 1, 1, 1]]), 
'labels': tensor([[16438,  6827,  -100,  -100,  -100,  -100,  -100],
                  [3844, 326, 338, 257, 2392, 6827, -100]])}
Notice how even though attention_mask properly has 1 for the EOS tokens, they still got set to -100.

Expected behavior
I'd expect that only the EOS tokens added as padding should be set to -100 in the labels, resulting in the following batch:

{'input_ids': tensor([[16438,  6827, 50256, 50256, 50256, 50256, 50256], 
                      [3844, 326, 338, 257, 2392, 6827, 50256]]), 
'attention_mask': tensor([[1, 1, 1, 0, 0, 0, 0], 
                          [1, 1, 1, 1, 1, 1, 1]]), 
'labels': tensor([[16438,  6827,  50256,  -100,  -100,  -100,  -100],
                  [3844, 326, 338, 257, 2392, 6827, 50256]])}
(notice the 50256 in labels)

I wanted to fine-tune GPT-2 for a rather specific use-case with very short texts, so I've added EOS to my samples and it took me quite a bit of time to spot what was the issue, causing the model to fail at generating short texts. After I spotted it, as a workaround I just set labels manually and use the standard collator instead, which does what I wanted.

I feel like this is a simple change [here](https://github.com/huggingface/transformers/blob/v4.24.0/src/transformers/data/data_collator.py#L738), to use the attention mask instead of [labels == self.tokenizer.pad_token_id].
I'd like to make a PR with that, but I just want to make sure that this is indeed a bug, and not expected behaviour.
